agenda/data/src/main/java/com/tasky/agenda/data/repositories/OfflineFirstAgendaItemsRepository.kt

```diff
@@ -116,4 +142,65 @@ class OfflineFirstAgendaItemsRepository(
         return result
     }
 
+    override suspend fun syncPendingAgendaItems() {
+        withContext(Dispatchers.IO) {
+            val userId = authInfoStorage.get()?.userId ?: return@withContext
+
+            eventSyncDao.getAllEventPendingSyncs(
+                syncType = SyncType.CREATE,
+                userId = userId
+            ).forEach {
+                eventSyncScheduler.sync(EventSyncScheduler.SyncType.CreateEventSync(it.event.toEvent()))
+            }
+            eventSyncDao.getAllEventPendingSyncs(
+                syncType = SyncType.UPDATE,
+                userId = userId
+            ).forEach {
+                eventSyncScheduler.sync(EventSyncScheduler.SyncType.UpdateEventSync(it.event.toEvent()))
+            }
+            eventDeleteSyncDao.getAllEventDeletePendingSyncs(
+                userId = userId
+            ).forEach {
+                eventSyncScheduler.sync(EventSyncScheduler.SyncType.DeleteEventSync(it.eventId))
+            }
```

Comment:
In this function, the database gets accessed multiple times in a synchronous manner. So before you can schedule the reminder sync tasks, first the events and the task events need to complete. To me, that does not make much sense - why then have these operations depend on each other? 🤔 

In a situation like this, we can use Kotlin’s `async` coroutine builder to have all our operations happen asynchronously. Check out this sample:

```
override suspend fun syncPendingAgendaItems() {
  withContext(Dispatchers.IO) {
    val userId = authInfoStorage.get()?.userId ?: return@withContext
    
    val createdEvents = async {
      eventSyncDao.getAllEventPendingSyncs(
        syncType = SyncType.CREATE,
        userId = userId
      )
    }
    
    val deletedEvents = async {
      eventDeleteSyncDao.getAllEventDeletePendingSyncs(userId)
    }
    
    // .. other sync calls
    
    val createEventJobs = createdEvents.await().map { pendingEntity ->
      launch {
        val event = pendingEntity.event.toDomain() // convert to domain model Event
        
        // try to make the API call immediately
        when(remoteDataSource.postEvent(event)) {
          is Result.Error -> Unit
          is Result.Success -> applicationScope.launch {
            pendingSyncDao.deleteEventPendingSyncEntity(event.id)
          }.join()
        }
      }
    }
    
    // … other API calls
    
    // Then we await the completion of each before the function terminates
    createEventJobs.forEach { it.join() }
    deleteEventJobs.forEach { it.join() }
    createTaskJobs.forEach { it.join() }
    deletedTaskJobs.forEach { it.join() }
    // ...    
  }
}
```